stew/base10.nim

```diff
@@ -0,0 +1,179 @@
+## Copyright (c) 2021 Status Research & Development GmbH
+## Licensed under either of
+##  * Apache License, version 2.0, ([LICENSE-APACHE](LICENSE-APACHE))
+##  * MIT license ([LICENSE-MIT](LICENSE-MIT))
+## at your option.
+## This file may not be copied, modified, or distributed except according to
+## those terms.
+
+## This module implements BASE10 (decimal) encoding and decoding procedures.
+##
+## Encoding procedures are adopted versions of C functions described here:
+## # https://www.facebook.com/notes/facebook-engineering/three-optimization-tips-for-c/10151361643253920
+import results
+export results
+
+{.push raises: [Defect].}
+
+type
+  Base10* = object
+
+{.push overflowChecks: off.}
+proc decode*[A: byte|char](B: typedesc[Base10], T: typedesc[SomeUnsignedInt],
+                           src: openarray[A]): Result[T, cstring] =
+  ## Convert base10 encoded string or array of bytes to unsigned integer.
+  const
+    MaxValue = T(high(T) div 10)
+    MaxNumber = T(high(T) - MaxValue * 10)
+
+  if len(src) == 0:
+    return err("Missing decimal value")
+  var v = T(0)
+  for i in 0 ..< len(src):
+    let ch = src[i]
+    let d =
+      when A is char:
+        if ch >= char(0x30'i8) and ch <= char(0x39'i8):
+          int(int8(ch) - 0x30'i8)
```

Comment:
the conversion to int here is no longer needed actually - we can go straight to T - without it, we can also remove the push overflowchecks off, and work with unsigned only. also, `char` is actually unsigned in nim so we can also `let ch = when A is char:  byte(src[i]) else: src[i]` which should further simplify the code